### PR TITLE
Plugins access control

### LINKED ISSUE #3328: Access control for plugins on cloud
(Copying/adapting from [this comment](https://github.com/PostHog/posthog/pull/3291#issuecomment-776960318))

For cloud we need to have some level of access control at the organization level. This could be done at the `env` level for now, though for ease of use we probably should to store this in the database on the `Organization`.

There will be four levels of access we need to grant plugins:

1. Global access. You can install plugins for your organisation and enable them per project. You can also toggle a boolean `global` on any installed plugin that will make it available for all organisations to use. On cloud our organisation (or a secure "plugin control" organisation) will have this access and use it to keep everyone's default plugins up to date.
2. Install access. You can install plugins for your organisation and enable them per project. I suspect we will eventually give this access to some EE customers on cloud, who will be installing their own custom plugins. They'll still have access to the globally installed plugins. 
3. Configuration access. You can enable installed plugins. This we would eventually turn on for everyone on cloud.
4. No plugin access, but still ingesting events via the plugin server. Default for new accounts on cloud for now.

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

Basically, resolves #3328. Context in the issue.

## Checklist

- [x] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests

### CODE REVIEW

**Comment by Twixes** [comment]
Here's how this works:
- The Plugin model now has an `is_global` bool field (default false), which can only be set by **root** orgs (see point below for that). When false, it's business as usual, installed only for this organization. However, when true, the plugin is marked as installed for **all organizations** in the PostHog instance.

- The Organization model now has a `plugins_access_level` int field (default 0 on Cloud, otherwise 9), where 0 is no access, 3 is config access, 6 is install access, 9 is root access:
    - **no access** can't see the Plugins page at all
    - **config access** can only use globally installed plugins managed by another org(s)
    - **install access** can use globally installed plugins, but can also install their own from npm/GitHub/GitLab/code editor
    - **root** has full capabilities, with the ability to manage plugins for all orgs

root vs. config (installed)

root vs. config (drawers)

File: posthog/api/plugin.py [hunk 1]
```diff
@@ -91,7 +106,15 @@ def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:
...
+        if (
+            validated_data.get("is_global")
+            and Organization.objects.only("plugins_access_level")
+            .get(id=self.context["organization_id"])
+            .plugins_access_level
+            < Organization.PluginsAccessLevel.ROOT
+        ):
+            raise PermissionError("This organization can't manage global plugins!")
+        if not validated_data and plugin.plugin_type != Plugin.PluginType.SOURCE:
             validated_data = self._update_validated_data_from_url({}, plugin.url)
```
Comment (mariusandra) [inline posthog/api/plugin.py:118]: We probably need to split this into two actions: 1) updating the is_global prop, 2) updating to the latest plugin version

File: frontend/src/scenes/plugins/pluginsLogic.ts [hunk 2]
```diff
@@ -110,11 +111,13 @@ export const pluginsLogic = kea<
...
-                updatePlugin: async ({ id }) => {
+                updatePlugin: async ({ id, pluginChanges }) => {
                     const { plugins } = values
-                    const response = await api.update(`api/organizations/@current/plugins/${id}`, {})
+                    const response = await api.update(`api/organizations/@current/plugins/${id}`, pluginChanges)
                     capturePluginEvent(`plugin updated`, response)
-                    actions.pluginUpdated(id)
+                    if (!Object.keys(pluginChanges).length) {
+                        actions.pluginUpdated(id)
+                    }
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/pluginsLogic.ts:120]: Small problem here. This endpoint by default fetches the latest version of the plugin. It, ughm, updates it, :). Sending `pluginChanges` here will make it a) update to the newest verison, b) update pluginChanges. I think we should split this into two actions. I definitely wouldn't expect a plugin to upgrade if I click the "is global" checkbox.

**Review by mariusandra — CHANGES_REQUESTED** [review]
Great progress and seems solid. Found still a few things to comment on below.

**Comment by mariusandra** [comment]
Tests fail as well :)

File: frontend/src/scenes/plugins/pluginsLogic.ts (hunk 2 above)
Comment (mariusandra) [inline frontend/src/scenes/plugins/pluginsLogic.ts:120]: I think it should somehow be more clear that if you call this function without pluginChanges, it will update the plugin to the latest available version... 🤔

**Comment by Twixes** [comment]
Oh, sorry, the code isn't all reviewable just yet, definitely deserves cleanup + tests. But the system is in place.

File: frontend/src/scenes/plugins/pluginsLogic.ts (hunk 2 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/pluginsLogic.ts:120]: Yes, TBH I'd just move plugin **version** update to a separate POST-only action endpoint, like: `.../plugins/:id/upgrade`. WDYT?

File: posthog/settings.py [hunk 3]
```diff
@@ -497,7 +497,7 @@ def print_warning(warning_lines: Sequence[str]):
 DEFAULT_FROM_EMAIL = os.getenv("EMAIL_DEFAULT_FROM", os.getenv("DEFAULT_FROM_EMAIL", "root@localhost"))
 EMAIL_REPLY_TO = os.getenv("EMAIL_REPLY_TO")
 
-MULTI_TENANCY = False  # overriden by posthog-cloud
+MULTI_TENANCY = True  # overriden by posthog-cloud
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/settings.py:500]: Yep, testing only, as this PR practically only affects Cloud functionality (though private instance admins can also downgrade orgs if they see fit).

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx [hunk 4]
```diff
@@ -157,6 +163,35 @@ export function PluginDrawer(): JSX.Element {
                                             <EnabledDisabledSwitch />
                                         </Form.Item>
                                     </div>
+                                    {user?.organization?.plugins_access_level === PluginsAccessLevel.Root && (
+                                        <div style={{ display: 'flex', alignItems: 'center' }}>
+                                            <Form.Item
+                                                fieldKey="__enabled"
+                                                name="__enabled"
+                                                style={{ display: 'inline-block', marginBottom: 0, color: red.primary }}
+                                            >
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/plugins/edit/PluginDrawer.tsx:172]: Good point, that looks like a better place

File: frontend/src/scenes/plugins/pluginsLogic.ts (hunk 2 above)
Comment (mariusandra) [reply frontend/src/scenes/plugins/pluginsLogic.ts:120]: There's actually one edge case with the current update that I already bumped into.

I might see this:

image

... but by the time I'm done checking the diff and click "update", a new version of the plugin is out. I won't upgrade to what's on the screen but something else. If I meticulously scan the code of a plugin between version upgrades, then I might miss something.

What actually happens is that the "latest version" tag will remain what it says, but "current version" (`tag`) will update to the next version... and we'll still see that there's a new version available.

Passing a full `url` with the right tag to the update command would solve this... and make it all easier to reason about.

**Comment by Twixes** [comment]
Up for review. I've added tests for various permissioning cases, feel free to inspect them and if anything seems daft (or unreadable), point it out, as this may have security implications.

File: frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx [hunk 5]
```diff
@@ -69,7 +70,9 @@ export function InstalledTab(): JSX.Element {
     } = useActions(pluginsLogic)
 
     const upgradeButton =
-        user?.plugin_access.install && hasNonSourcePlugins ? (
+        user?.organization &&
+        user.organization.plugins_access_level === PluginsAccessLevel.Root &&
+        hasNonSourcePlugins ? (
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx:75]: If you have install access, you should still be able to upgrade your own plugins. The individual upgrade buttons still show now...

File: posthog/migrations/0132_add_plugins_updated_created_at.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 3.0.6 on 2021-02-25 16:03
```
Comment (mariusandra) [inline posthog/migrations/0132_add_plugins_updated_created_at.py:1]: The filename of this migration is a bit off

File: frontend/src/scenes/plugins/Plugins.tsx [hunk 7]
```diff
@@ -49,14 +50,16 @@ function _Plugins(): JSX.Element {
...
-                    {user.plugin_access.install ? (
+                    {user.organization?.plugins_access_level >= PluginsAccessLevel.Install ? (
                         <Tabs activeKey={pluginTab} onChange={(activeKey) => setPluginTab(activeKey as PluginTab)}>
                             <TabPane tab="Installed" key={PluginTab.Installed}>
                                 <InstalledTab />
                             </TabPane>
-                            <TabPane tab="Repository" key={PluginTab.Repository}>
-                                <RepositoryTab />
-                            </TabPane>
+                            {user.organization?.plugins_access_level === PluginsAccessLevel.Root && (
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/Plugins.tsx:58]: Should be install level

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx [hunk 8]
```diff
@@ -172,6 +174,33 @@ export function PluginDrawer(): JSX.Element {
...
+                                    >
+                                        <Checkbox
+                                            checked={editingPlugin.is_global}
+                                            onChange={(e) =>
+                                                patchPlugin(editingPlugin.id, {
+                                                    is_global: e.target.checked,
+                                                })
+                                            }
+                                        />
+                                        <span style={{ paddingLeft: 10 }}>Mark as global</span>
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/edit/PluginDrawer.tsx:199]: Shouldn't this be inside the `<Checkbox>` tag? It's in any case not clickable.

I also find it a bit confusing to have a checkbox that autosaves the form, though all things considered, it's probably a fine compromise.

image

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx [hunk 9]
```diff
@@ -44,12 +46,17 @@ const SecretFieldIcon = (): JSX.Element => (
...
+        savePluginConfig,
+        uninstallPlugin,
+        setEditingSource,
+        generateApiKeysIfNeeded,
+        patchPlugin,
+    } = useActions(pluginsLogic)
     const [form] = Form.useForm()
 
-    const canDelete = user?.plugin_access.install
+    const canDelete = (user?.organization?.plugins_access_level ?? 0) >= PluginsAccessLevel.Install
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/edit/PluginDrawer.tsx:59]: Will the delete button show for plugins of other organisations that are global?

File: frontend/src/scenes/plugins/plugin/PluginCard.tsx [hunk 10]
```diff
@@ -127,8 +128,12 @@ export function PluginCard({
...
                                 <PluginError error={error} />
                             ) : null}
-
-                            {canInstall ? (
+                            {is_global && (
+                                <Tag color="blue">
+                                    <GlobalOutlined /> Managed by {organization_name}
+                                </Tag>
+                            )}
+                            {user?.organization?.plugins_access_level === PluginsAccessLevel.Root && (
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/plugin/PluginCard.tsx:136]: These shouldn't be hidden by access level alone, but also shown for the install level if it's a plugin for their organisation.

File: frontend/src/scenes/plugins/pluginsLogic.ts [hunk 11]
```diff
@@ -492,11 +510,14 @@ export const pluginsLogic = kea<
...
             }
-            if (userLogic.values.user?.plugin_access.install) {
+            if ((userLogic.values.user?.organization?.plugins_access_level ?? 0) >= PluginsAccessLevel.Install) {
                 actions.checkForUpdates(false, initialUpdateStatus)
-            }
-            if (Object.keys(values.plugins).length === 0 && userLogic.values.user?.plugin_access.install) {
-                actions.setPluginTab(PluginTab.Repository)
+                if (
+                    Object.keys(values.plugins).length === 0 &&
+                    organizationLogic.values.currentOrganization?.plugins_access_level === PluginsAccessLevel.Root
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/pluginsLogic.ts:517]: Also for install access

File: frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx [hunk 12]
```diff
@@ -222,9 +226,11 @@ export function InstalledTab(): JSX.Element {
                         <Row gutter={16} style={{ marginTop: 16 }}>
                             <Col span={24}>
                                 <Empty description={<span>You haven't installed any plugins yet</span>}>
-                                    <Button type="default" onClick={() => setPluginTab(PluginTab.Repository)}>
-                                        Open the Plugin Repository
-                                    </Button>
+                                    {user?.organization?.plugins_access_level === PluginsAccessLevel.Root && (
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/tabs/installed/InstalledTab.tsx:229]: also for install

**Review by mariusandra — CHANGES_REQUESTED** [review]
I went through the frontend code in depth and played with it locally. I'll go through the python code as well once the following is cleared up.

Basically, I'm not sure I understand well the intention between splitting the access levels for cloud and self hosted with `install` vs `root`. Definitely `root` is needed to toggle the `is_global` flag, but that should be it, right? Cloud users with `install` rights should still be able to install and update their own plugins (and not delete/update plugins of other organisations, global or otherwise).

The levels should be:
- Configure - can enable/disable/rearrange plugins and edit their config
- Install - can install any plugin or write your own. Self-Hosted permissions stop here (there's just 1 org, so the next level makes no sense)
- Admin/Root/Superuser/James - can also set plugins to be "global" and share them between orgs.

There were some interface access right bugs as well. For example, running with the `Install` access and with or without `MULTI_TENANCY`, I was able to upgrade and delete plugins in the interface, but all tools to add new plugins were hidden.

I'll check the Python part once the access levels are cleared up.

**Comment by Twixes** [comment]
Alright, so I think we're on the same page with levels none and config. As for the other two:

`INSTALL` - that is is only for Cloud usage, for paid orgs we trust (eventually all orgs). Maybe the name isn't precise enough, because it is **not intended for full installation capabilities**. Specifically, the repository is off limits, as it's assumed that repository plugins are managed by root org(s) – allowing duplicated installations of plugins from the repository would only be confusing. Particularly on Cloud, but otherwise too.

`ROOT` - this is intentionally the default for self-hosted orgs. No restrictions – they run the instance after all. And if they want to have multiple organizations on the same instance for whatever reason, the instance admin can manage plugins for all orgs as an option.

What would you see differently regarding these?

**Comment by Twixes** [comment]
As far as I understood you'd prefer if `INSTALL` on self-hosted had capabilities just like `ROOT`, only without the global management option. I didn't go that way, because the plugins repository situation makes `ROOT` much closer to the intended effect on self-hosted. I think we can hide the "Manage globally" checkbox on self-hosted as I can see it being more confusing for our self-hosting users than really useful, but making `INSTALL` and `ROOT` more conditional on `settings.MULTI_TENANCY` would only increase complexity.

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx (hunk 8 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/edit/PluginDrawer.tsx:199]: Yep, should be clickable.

This does bypass "Save" which is not ideal UX, but unfortunately it has to, as this is a control for the Plugin model, while the rest of the drawer is a control for PluginConfig. Different endpoints.
I actually wanted to make this a button next to "Uninstall" (which does correspond to the Plugin model), but unfortunately there's not enough space in the footer.

File: frontend/src/scenes/plugins/Plugins.tsx (hunk 7 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/Plugins.tsx:58]: Rationale in comment.

Comment (mariusandra) [reply frontend/src/scenes/plugins/Plugins.tsx:58]: well, but this disables it also for self hosted users :)

**Comment by mariusandra** [comment]
I see... just a bit of context from the deviation from the issue and current state would have helped. I was going into the review knowing the existing meaning of `install`... and without knowing what might now seem obvious design choices that came out while coding the thing. :)

The proposed changes make sense. OK to have `ROOT` be the default for self-hosted... and to hiding the repository for `INSTALL` aka "cloud users". Just I'd somehow hide the "Mark as global" checkbox if there's only one organisation. Otherwise it feels like it would be a confusing checkbox. (Or add a enterprise version upsell modal 🤑)

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx (hunk 9 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/edit/PluginDrawer.tsx]: Not anymore

**Comment by Twixes** [comment]
Addressed UX feedback (commit messages 👆 should be conveying details). You're right, I should've laid out the specifics of the system better – added comments to `Organization.PluginsAccessLevel` with some relevant info.

File: frontend/src/scenes/plugins/plugin/PluginCard.tsx [hunk 13]
```diff
@@ -130,7 +130,7 @@ export function PluginCard({
                                     <GlobalOutlined /> Managed by {organization_name}
                                 </Tag>
                             )}
-                            {user?.organization?.plugins_access_level === PluginsAccessLevel.Root && (
+                            {user?.organization?.id === organization_id && (
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/plugin/PluginCard.tsx:133]: Shouldn't this still check the access level as well?

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx [hunk 14]
```diff
@@ -95,7 +93,7 @@ export function PluginDrawer(): JSX.Element {
                     <>
                         <div style={{ display: 'flex' }}>
                             <div style={{ flexGrow: 1 }}>
-                                {canDelete && (
+                                {editingPlugin?.organization_id === user?.organization?.id && (
```
Comment (mariusandra) [inline frontend/src/scenes/plugins/edit/PluginDrawer.tsx:96]: Might be wise to show this only for `INSTALL` and greater?

**Review by mariusandra — COMMENTED** [review]
Two extra frontend comments for now...

File: frontend/src/scenes/plugins/plugin/PluginCard.tsx (hunk 13 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/plugin/PluginCard.tsx:133]: It doesn't have to, `organization_id` has to be from an org which has at least install access by definition.

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx (hunk 14 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/edit/PluginDrawer.tsx]: Alright, it's very much an edge case like the other comment, but added access level checks to both places.

**Comment by mariusandra** [comment]
@macobo I'm unfortunately going to have to put this on your lap as well. I have reviewed the frontend and all seems good there! Yet I did not manage to go through any `.py` file yet. I'm not that at home with permissions in django, and hence have to be really meticulous when reviewing this and test everything... but don't have the time anymore. :/

File: frontend/src/lib/utils.tsx [hunk 15]
```diff
@@ -747,3 +747,14 @@ export function sortedKeys(object: Record<string, any>): Record<string, any> {
     }
     return newObject
 }
+
+export function endWithPeriod(text?: string | null): string {
```
Comment (macobo) [inline frontend/src/lib/utils.tsx:751]: Suggestion: `endWithPunctation`.

File: frontend/src/scenes/plugins/Plugins.tsx [hunk 16]
```diff
@@ -47,14 +48,16 @@ export function Plugins(): JSX.Element {
...
```
Comment (macobo) [inline frontend/src/scenes/plugins/Plugins.tsx:56]: This code is not self-evident, it's not clear to a reader what `plugin_access_level` stands for or what the right levels are.

Suggestion:

Extract module `frontend/src/scenes/plugins/accessControl.ts` containing function(s) like `hasAccessToPlugins(user)`,  `canInstallPlugins(user)` etc. This keeps the business logic centralized and the component files readable.

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx [hunk 17]
```diff
@@ -170,6 +173,35 @@ export function PluginDrawer(): JSX.Element {
                                 </div>
                             ) : null}
 
+                            {user?.organization?.plugins_access_level === PluginsAccessLevel.Root &&
+                                user?.is_multi_tenancy && (
```
Comment (macobo) [inline frontend/src/scenes/plugins/edit/PluginDrawer.tsx:177]: Q: Will this mean that this is a cloud user or...?

Also: Extract function `canEnableGlobalPlugins`.

File: frontend/src/scenes/plugins/optin/OptInPlugins.tsx [hunk 18]
```diff
@@ -15,8 +15,8 @@ export function OptInPlugins(): JSX.Element {
     useEffect(() => {
         async function setStatus(): Promise<void> {
             try {
-                const response = await api.get('api/organizations/@current/plugins/status')
```
Comment (macobo) [inline frontend/src/scenes/plugins/optin/OptInPlugins.tsx:18]: Q: Context for this change?

File: frontend/src/scenes/plugins/plugin/PluginCard.tsx [hunk 19]
```diff
@@ -124,35 +126,40 @@ export function PluginCard({
...
-                                    ) : url && latestTag && tag ? (
-                                        tag === latestTag ? (
-                                            <Tag color="green">
-                                                <CheckOutlined /> Up to date
+                            {is_global && (
+                                <Tag color="blue">
+                                    <GlobalOutlined /> Managed by {organization_name}
+                                </Tag>
+                            )}
+                            {user?.organization?.id === organization_id &&
```
Comment (macobo) [inline frontend/src/scenes/plugins/plugin/PluginCard.tsx:134]: Suggestion: `canManage(user, plugin)`

File: posthog/api/plugin.py [hunk 20]
```diff
@@ -27,15 +28,36 @@
...
 SECRET_FIELD_VALUE = "**************** POSTHOG SECRET FIELD ****************"
 
 
+class PluginsAccessLevelPermission(BasePermission):
+    message = "Your organization's plugin access level is insufficient."
+
+    def has_permission(self, request, view) -> bool:
+        min_level = (
+            Organization.PluginsAccessLevel.CONFIG
+            if request.method in SAFE_METHODS
```
Comment (macobo) [inline posthog/api/plugin.py:45]: Q: Where does SAFE_METHODS come from?

File: posthog/api/plugin.py [hunk 21]
```diff
@@ -89,16 +119,27 @@ def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:
...
 
     def update(self, plugin: Plugin, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:  # type: ignore
-        if not can_install_plugins_via_api(self.context["organization_id"]):
-            raise ValidationError("Plugin upgrades via the web are disabled!")
-        if plugin.plugin_type != Plugin.PluginType.SOURCE:
-            validated_data = self._update_validated_data_from_url({}, plugin.url)
+        context_organization = self.context.get("organization") or Organization.objects.get(
+            id=self.context["organization_id"]
+        )
+        if plugin.organization != context_organization:
```
Comment (macobo) [inline posthog/api/plugin.py:125]: Q: is this not duplicating PluginOwnershipPermission and PluginsAccessLevelPermission? Why?

File: posthog/models/plugin.py [hunk 22]
```diff
@@ -15,6 +15,7 @@ class PluginType(models.TextChoices):
     plugin_type: models.CharField = models.CharField(
         max_length=200, null=True, blank=True, choices=PluginType.choices, default=None
     )
+    is_global: models.BooleanField = models.BooleanField(default=False)  # Whether plugin is installed for all orgs
```
Comment (macobo) [inline posthog/models/plugin.py:18]: Q: Will merging this break plugin server tests again (due to extra cols)?

File: posthog/models/team.py [hunk 23]
```diff
@@ -88,7 +88,9 @@ class Team(models.Model):
     ingested_event: models.BooleanField = models.BooleanField(default=False)
     uuid: models.UUIDField = models.UUIDField(default=UUIDT, editable=False, unique=True)
     session_recording_opt_in: models.BooleanField = models.BooleanField(default=False)
-    session_recording_retention_period_days: models.IntegerField = models.IntegerField(null=True, default=None)
+    session_recording_retention_period_days: models.IntegerField = models.IntegerField(
+        null=True, default=None, blank=True
```
Comment (macobo) [inline posthog/models/team.py:92]: Unrelated change? Also doesn't blank mean empty string? This value is an integer so this seems wrong.

File: posthog/plugins/utils.py [hunk 24]
```diff
@@ -42,12 +42,14 @@ def parse_github_url(url: str, get_latest_if_none=False) -> Optional[Dict[str, O
...
             commits = requests.get(commits_url, headers=headers).json()
+            if isinstance(commits, dict):
+                raise Exception(commits.get("message"))
             if len(commits) > 0 and commits[0].get("sha", None):
                 parsed["tag"] = commits[0]["sha"]
             else:
-                raise
-        except Exception:
-            raise Exception("Could not get latest commit for: {}".format(parsed["root_url"]))
+                raise Exception()
```
Comment (macobo) [inline posthog/plugins/utils.py:50]: Add a message here?

File: posthog/settings.py [hunk 25]
```diff
@@ -71,11 +71,6 @@ def print_warning(warning_lines: Sequence[str]):
 else:
     JS_URL = os.getenv("JS_URL", "")
 
-PLUGINS_INSTALL_VIA_API = get_from_env("PLUGINS_INSTALL_VIA_API", True, type_cast=strtobool)
-PLUGINS_CONFIGURE_VIA_API = PLUGINS_INSTALL_VIA_API or get_from_env(
```
Comment (macobo) [inline posthog/settings.py:75]: Q: Context on removing these?

File: posthog/plugins/access.py [hunk 26]
```diff
@@ -3,28 +3,52 @@
...
 
 
-# We disable Plugins on Cloud, except for whitelisted organizations
-# Disregarding this in TEST mode, so that we can be sure plugins actually work in EE if/when needed
-def guard_cloud(organization_or_id: Optional[Union[Organization, UUID, str]]):
-    organization_id: Optional[str] = (
-        None
-        if not organization_or_id
-        else str(organization_or_id if isinstance(organization_or_id, (str, UUID)) else organization_or_id.id)
+def can_root_plugins_via_api(
```
Comment (macobo) [inline posthog/plugins/access.py:9]: This file does not seem to have any unit tests

File: posthog/api/plugin.py [hunk 27]
```diff
@@ -181,6 +226,22 @@ def check_for_updates(self, request: request.Request, **kwargs):
 
         return Response({"plugin": PluginSerializer(plugin).data})
 
+    @action(methods=["POST"], detail=True)
+    def upgrade(self, request: request.Request, **kwargs):
```
Comment (macobo) [inline posthog/api/plugin.py:230]: No unit tests?

File: posthog/api/plugin.py (hunk 21 above)
Comment (macobo) [inline posthog/api/plugin.py:125]: Q: is this not duplicating PluginOwnershipPermission and PluginsAccessLevelPermission? Why?

File: posthog/api/plugin.py [hunk 28]
```diff
@@ -89,16 +119,27 @@ def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:
         return plugin
 
     def update(self, plugin: Plugin, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:  # type: ignore
```
Comment (macobo) [inline posthog/api/plugin.py:121]: No tests for changes?

**Review by macobo — CHANGES_REQUESTED** [review]
Did my first code-level pass. The logic looks great, but I think the FE code can be cleaned up a lot. For the backend, seeing an absence of tests?

File: posthog/api/plugin.py (hunk 28 above)
Comment (Twixes) [reply posthog/api/plugin.py:121]: WDYM?

Comment (macobo) [reply posthog/api/plugin.py:121]: I might have missed this, but found no tests for the changes in this file.

File: posthog/settings.py (hunk 25 above)
Comment (Twixes) [reply posthog/settings.py:75]: We've never actually used these. And now we are moving to an approach based on `Organization.plugins_access_level`, which will actually be used.

File: posthog/api/plugin.py (hunk 28 above)
Comment (Twixes) [reply posthog/api/plugin.py:121]: Actually I most of `test_plugin.py` is red or green. Indeed the diff is so big the file has become a victim of GitHub's "Large diffs are not rendered by default." so that's probably why you missed it.

Comment (macobo) [reply posthog/api/plugin.py:121]: :D Sorry for the misleading comment then!

File: frontend/src/lib/utils.tsx (hunk 15 above)
Comment (Twixes) [reply frontend/src/lib/utils.tsx:751]: Why not

File: frontend/src/scenes/plugins/Plugins.tsx (hunk 16 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/Plugins.tsx]: Done

File: frontend/src/scenes/plugins/optin/OptInPlugins.tsx (hunk 18 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/optin/OptInPlugins.tsx:18]: Deduplicating the API. Otherwise we had two endpoints doing the same thing slightly differently.

File: posthog/api/plugin.py (hunk 20 above)
Comment (Twixes) [reply posthog/api/plugin.py:45]: It's a DRF thing, basically `['GET', 'HEAD', 'OPTIONS']`.

File: posthog/api/plugin.py (hunk 21 above)
Comment (Twixes) [reply posthog/api/plugin.py:125]: It was, holdover. Deduplicated.

File: posthog/models/plugin.py (hunk 22 above)
Comment (Twixes) [reply posthog/models/plugin.py:18]: It will, https://github.com/PostHog/plugin-server/pull/230 standing by

File: posthog/models/team.py (hunk 23 above)
Comment (Twixes) [reply posthog/models/team.py:92]: It's only a validation thing, we weren't able to edit organizations in Django admin otherwise. In fact blank is not string related at all, it only means that the field can be left empty (in this case null). Otherwise Django admin wouldn't save the org edit without providing a number in this field. The same applies to these Team fields updated here.

File: posthog/api/plugin.py (hunk 27 above)
Comment (Twixes) [reply posthog/api/plugin.py:230]: `test_plugin.py`

File: posthog/plugins/access.py (hunk 26 above)
Comment (Twixes) [reply posthog/plugins/access.py:9]: Added

File: frontend/src/scenes/plugins/edit/PluginDrawer.tsx (hunk 17 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/edit/PluginDrawer.tsx]: Didn't extract the `is_multi_tenancy` since it'd require passing in `user` too and it's not globally relevant to permissioning. Just a presentation thing for now. Did extract the Root (etc.) check though.

File: frontend/src/scenes/plugins/pluginsLogic.ts [hunk 29]
```diff
@@ -112,11 +114,9 @@ export const pluginsLogic = kea<
                     return { ...plugins, [id]: response }
                 },
                 updatePlugin: async ({ id }) => {
-                    const { plugins } = values
-                    const response = await api.update(`api/organizations/@current/plugins/${id}`, {})
+                    const response = await api.create(`api/organizations/@current/plugins/${id}/upgrade`)
```
Comment (macobo) [inline frontend/src/scenes/plugins/pluginsLogic.ts:117]: Note: I find the update/upgrade vs patch semantics a bit weird, but not sure how to improve this.

**Review by macobo — APPROVED** [review]
Good work. :shipit:

File: frontend/src/scenes/plugins/pluginsLogic.ts (hunk 29 above)
Comment (Twixes) [reply frontend/src/scenes/plugins/pluginsLogic.ts:117]: (blame Marius) Well initially `PATCH` was responsible for triggering this, but considering `PATCH` is represented as `update` in our API helper and in DRF, it is confusing. Had to change the endpoint to a separate `/upgrade`, but in the UI the feature is still "update".

Comment (macobo) [reply frontend/src/scenes/plugins/pluginsLogic.ts:117]: Agreed. This is non-blocking imo.
